pkg/ottl/grammar.go

```diff
@@ -121,9 +183,35 @@ type comparison struct {
 	Right value     `parser:"@@"`
 }
 
-// invocation represents a function call.
+func (c *comparison) checkForCustomError() error {
+	err := c.Left.checkForCustomError()
+	if err != nil {
+		return err
+	}
+	err = c.Right.checkForCustomError()
+	return err
+}
+
+// invocation represents the function call of a statement.
 type invocation struct {
-	Function  string  `parser:"@(Uppercase | Lowercase)+"`
+	Function  string  `parser:"@(Lowercase(Uppercase | Lowercase)*)"`
+	Arguments []value `parser:"'(' ( @@ ( ',' @@ )* )? ')'"`
+}
+
+func (i *invocation) checkForCustomError() error {
+	var err error
+	for _, arg := range i.Arguments {
+		err = arg.checkForCustomError()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// converter represents a converter function call.
+type converter struct {
+	Function  string  `parser:"@(Uppercase(Uppercase | Lowercase)*)"`
```

Comment:
Sorry I just noticed this, but it means converter and invocation names cannot contain digits or underscores. This will preclude names like Atan2 and do_something. I think both things should probably allow for the 2-nth characters to be in `[A-Za-z0-9_]*`.

Reply:
@kentquirk the grammar actually allows the `Uppercase` and `Lowercase` tokens to include underscore as a starting character and then a mix of the correct case characters, numbers and underscores as following characters

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/0af9ff5fa46085a3acbcbd531742b034df8f28ee/pkg/ottl/grammar.go#L280-L281

So we can already do `Atan2` and `do_something`. 

But I am seeing now we can do stuff like just `_` and that would break things I think.  Even `_wrong` might incorrectly match to `"@(Uppercase(Uppercase | Lowercase)*)"`.  

I'm ok removing the leading underscore entirely and then we'll be good.

Reply:
Suggest adding names to the grammar for ConverterName as `[A-Z][A-Z0-9_]*` and then InvocationName as `[a-z][A-Z0-9_]*`, then using just `parser:"@ConverterName"` here.

Reply:
@kentquirk and I had a quick chat about this.

Due to path and enum requiring all upper and all lowercase letters we can't get away with a rename and then using exactly 1 token.  With the simple lexer we're using I believe this is the only way in the lexer to force paths to be lowercase, enums to be uppercase, functions to start with a lowercase name, and converters to start with an uppercase name.

One thing we could do in a separate PR though is to go back to the original Ident token and then use thar participle package's `Capture` interface to do specific capturing for path, enum, function name, and converter name.  In the capture function we should be able to check the incoming string for validity.